src/main/java/net/floodlightcontroller/forwarding/Forwarding.java

```diff
@@ -1088,4 +1669,234 @@ public void linkDiscoveryUpdate(List<LDUpdate> updateList) {
 
     }
 
+    @Override
+    public Collection<VirtualGatewayInstance> getGatewayInstances() { return l3manager.getAllVirtualGateways(); }
+
+    @Override
+    public Optional<VirtualGatewayInstance> getGatewayInstance(String name) { return l3manager.getVirtualGateway(name); }
+
+    @Override
+    public Optional<VirtualGatewayInstance> getGatewayInstance(DatapathId dpid) {
+        return l3manager.getVirtualGateway(dpid);
+    }
+
+    @Override
+    public Optional<VirtualGatewayInstance> getGatewayInstance(NodePortTuple npt) {
+        return l3manager.getVirtualGateway(npt);
+    }
+
+    @Override
+    public void deleteGatewayInstances() {
+        log.info("All virtual gateways deleted");
+        l3manager.removeAllVirtualGateways();
+    }
+
+    @Override
+    public boolean deleteGatewayInstance(String name) {
+        log.info("Virtual gateway {} deleted", name);
+        return l3manager.removeVirtualGateway(name);
+    }
+
+    @Override
+    public void addGatewayInstance(VirtualGatewayInstance gateway) {
+        log.info("A new virtual gateway {} created", gateway.getName());
+        l3manager.addVirtualGateway(gateway);
+    }
+
+    @Override
+    public VirtualGatewayInstance updateVirtualGateway(String name, MacAddress newMac) {
+        log.info("Virtual gateway {} updated", name);
+        return l3manager.updateVirtualGateway(name, newMac);
+    }
+
+    @Override
+    public Optional<Collection<VirtualGatewayInterface>> getGatewayInterfaces(VirtualGatewayInstance gateway) {
```

Comment:
You want to avoid returning an optional collection. Just return an empty collection

Reply:
That's correct, shouldn't use Optional collection. Fixed it.